pkg/repository/v1/payloadstore.go

```diff
@@ -0,0 +1,482 @@
+		}
+
+		return nil
+	case PayloadLocationExternal:
+		if p.ExternalLocationKey == nil || len(*p.ExternalLocationKey) == 0 {
+			return fmt.Errorf("external location key cannot be empty when location is %s", PayloadLocationExternal)
+		}
+
+		if len(p.InlineContent) > 0 {
+			return fmt.Errorf("inline content must be empty when location is %s", PayloadLocationExternal)
+		}
+
+		return nil
+	default:
+		return fmt.Errorf("invalid payload location: %s", p.Location)
+	}
+}
+
+func (p PayloadContent) Marshal() ([]byte, error) {
+	if err := p.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid payload: %w", err)
+	}
+
+	j, err := json.Marshal(p)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal payload: %w", err)
+	}
+
+	return j, nil
+}
+
+func (p *payloadStoreRepositoryImpl) unmarshalPayloadContent(data []byte) (*PayloadContent, error) {
+	var content PayloadContent
+	if err := json.Unmarshal(data, &content); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal payload content: %w", err)
+	}
+
+	if err := content.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid payload content: %w", err)
+	}
+
+	return &content, nil
+}
+
+func (p *payloadStoreRepositoryImpl) Store(ctx context.Context, tx sqlcv1.DBTX, payloads ...StorePayloadOpts) error {
+	taskIds := make([]int64, len(payloads))
+	taskInsertedAts := make([]pgtype.Timestamptz, len(payloads))
+	payloadTypes := make([]string, len(payloads))
+	payloadData := make([][]byte, len(payloads))
+	offloadAts := make([]pgtype.Timestamptz, len(payloads))
+	operations := make([]string, len(payloads))
+	tenantIds := make([]pgtype.UUID, len(payloads))
+
+	for i, payload := range payloads {
+		taskIds[i] = payload.Id
+		taskInsertedAts[i] = payload.InsertedAt
+		payloadTypes[i] = string(payload.Type)
+		tenantIds[i] = sqlchelpers.UUIDFromStr(payload.TenantId)
+
+		if p.externalStoreEnabled {
+			offloadAts[i] = pgtype.Timestamptz{Time: payload.InsertedAt.Time.Add(*p.nativeStoreTTL), Valid: true}
+		} else {
+			offloadAts[i] = pgtype.Timestamptz{Time: time.Now(), Valid: true}
+		}
+
+		operations[i] = string(sqlcv1.V1PayloadWalOperationCREATE)
+
+		// Always store inline initially - offloading happens via cron job
+		content := PayloadContent{
+			Location:      PayloadLocationInline,
+			InlineContent: payload.Payload,
+		}
+		marshaledContent, err := content.Marshal()
+
+		if err != nil {
+			return fmt.Errorf("failed to marshal inline payload: %w", err)
+		}
+
+		payloadData[i] = marshaledContent
+	}
+
+	err := p.queries.WritePayloads(ctx, p.pool, sqlcv1.WritePayloadsParams{
+		Tenantids:   tenantIds,
+		Ids:         taskIds,
+		Insertedats: taskInsertedAts,
+		Types:       payloadTypes,
+		Payloads:    payloadData,
+	})
+
+	if err != nil {
+		return fmt.Errorf("failed to write payloads: %w", err)
+	}
+
+	// only need to write to the WAL if we have an external store configured
+	if p.externalStoreEnabled {
```

Comment:
I'm a bit nervous about this short circuiting - feels like it'll make this harder to maintain in the future, but I also didn't want to have a ton of WAL overhead if it's just a no-op